docs/framework-docs.md

```diff
@@ -0,0 +1,327 @@
+
+There are currently 3 types of calls associated with user actions:
+- Submit - submit a form/command or click on a UI binding
+- Form - Fetch a form’s definition like a command or modal
+- Lookup - Fetch autocomplete results for am autocomplete form field
+
+A Call request is sent to the App’s server when:
+- The user visits a channel (call to fetch bindings)
+- The user clicks on a post menu or channel binding (may open a modal)
+- Commands:
+  - The user is filling out a command argument that fetches dynamic results (lookup call is performed)
+  - The user submits a command (may open a modal)
+- Modals:
+  - The user types a search in a modal’s autocomplete select field (lookup call is performed)
+  - The user selects a value from a “refresh” select element in a modal (the modal’s form will be re-fetched based on all filled out values)
+  - The user submits the modal (a new form may be returned from the App)
+- (TBD) A subscribed event like MessageHasBeenPosted occurs
+- (TBD) A third-party webhook request comes in
+
+
+## Context Expansion
+
+By default, only the IDs of certain entities are provided in the Call. Since we want these App functions to be stateless and quick, the App can specify certain objects to be "expanded" within the request to the App.
+
+The framework verifies that the user has access to the entities before expanding and including them in the request. Choosing which level of access token to use for expand depends on what the App initially requests when installed:
+
+- Bot token (possibly less privileged than admin)
+- Acting user token
+- Admin token
+
+
+Entities that can be expanded:
+
+- App - details about the App itself
+  - see what is provided in summary vs full
+
+- Acting User - model.User struct
+
+- ActingUserAccessToken - OAuth2 token for acting user
+  - currently is just acting MM session token
+
+- AdminAccessToken - Access token to do admin-level operations
+  - currently is just acting MM session token
+  - Eventually make a short-lived bot admin token, scoped to certain operations
+
+- Channel - The current channel the acting user is interacting with
+
+- Mentioned - Users/channels mentioned in the related post or command being run
+  - not implemented yet
+  - should be access controlled as far as what is able to be expanded per user/context
+
+- ParentPost - Parent post of the selected post
+- Post - Selected post, if the call is specific to a given post
+- RootPost - Root post of the selected post
+
+- Team - The current ream the acting user is interacting with
+
+
+### Conext Expansion Example
+
+In this example, the bindings specify to expand the post the user clicks on:
+
+![binding-form-diagram.png](https://user-images.githubusercontent.com/6913320/109165112-2e6ac800-7749-11eb-8d83-d495258f3f1e.png)
```

Comment:
@larkox You said you were working on sequence diagrams. Is this something similar to what you are working on?

Reply:
Exactly that. Here are some that I did, but they are mainly "brainstorming".
https://docs.google.com/document/d/18fhHqLboWWQiMzY_LJ91PlK6Tw_ziapzpUmABajPSxY/edit?usp=sharing